### PR TITLE
Feature/base repo

### PR DESCRIPTION
**Pull request by BenBoonya**

### CODE REVIEW

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            override fun onFailure(call: Call<T>?, t: Throwable?) {
+                failureHandler(null, null)
+                if (t is Exception) {
+                    t.printStackTrace()
+                }
+            }
+        })
+    }
+
+    protected fun <T> getError(response: Response<T>): ErrorResponse? {
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt]: Create Response extensions file

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        try {
+            val errorBody = response.errorBody().string()
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt]: Check optional

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt [hunk 3]
```diff
@@ -3,12 +3,21 @@ package com.ben.boonya.architecturecomponentdemo.base
...
 import android.arch.lifecycle.MediatorLiveData
+import android.arch.lifecycle.MutableLiveData
+import com.ben.boonya.architecturecomponentdemo.common.ErrorResponse
 
 /**
  * Created by oozou on 6/28/2017 AD.
  */
 open class BaseViewModel(application: Application) : AndroidViewModel(application) {
 
     val isLoading = MediatorLiveData<Boolean>()
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt]: MutableLiveData

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/Error.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import android.os.Parcel
+import android.os.Parcelable
+
+/**
+ * ErrorCrate
+ * on 7/27/2016
+
+ * @author Jutikorn Varojananulux <jutikorn.v></jutikorn.v>@gmail.com>
+ */
+class Error : Parcelable {
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/Error.kt]: data class

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/Error.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    var type = "N/A"
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/Error.kt]: @SerializeNamed

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/ErrorResponse.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import android.os.Parcelable
+
+
+/**
+ * ErrorResponse
+ * on 7/27/2016
+
+ * @author Jutikorn Varojananulux <jutikorn.v></jutikorn.v>@gmail.com>
+ */
+class ErrorResponse : Parcelable {
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/ErrorResponse.kt]: Data class

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/ErrorResponse.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    var error: Error? = null
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/ErrorResponse.kt]: @SerializeNamed

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt [hunk 8]
```diff
@@ -3,12 +3,21 @@ package com.ben.boonya.architecturecomponentdemo.base
...
-    val throwable = MediatorLiveData<Throwable>()
+    val apiError = MutableLiveData<Pair<ErrorResponse?, Int?>>()
+
+    fun handleError(errorResponse: ErrorResponse?, code: Int?) {
+        kotlin.run {
```
Comment (jutikorn) [inline app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt]: @BenBoonya check more what is kotlin.run

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt (hunk 3 above)
Comment (BenBoonya) [reply app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt]: Done

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/ErrorResponse.kt (hunk 6 above)
Comment (BenBoonya) [reply app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/ErrorResponse.kt]: done

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/Error.kt (hunk 4 above)
Comment (BenBoonya) [reply app/src/main/java/com/ben/boonya/architecturecomponentdemo/common/Error.kt]: done

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt (hunk 3 above)
Comment (BenBoonya) [reply app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseViewModel.kt]: done

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt (hunk 2 above)
Comment (BenBoonya) [reply app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt]: done

File: app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt (hunk 1 above)
Comment (BenBoonya) [reply app/src/main/java/com/ben/boonya/architecturecomponentdemo/base/BaseRepository.kt]: done

**Pull request merged**